File: Queue/Easy/first_negative_in_k_window.py
# https://practice.geeksforgeeks.org/problems/first-negative-integer-in-every-window-of-size-k3345/1#
#User function Template for python3

def printFirstNegativeInteger( A, N, k):
    q = []
    ans = []
    
    for i in range(k):
        if(A[i] < 0):
            q.append(i)
    
    if(len(q)> 0):
        ans.append(A[q[0]])
    else:
        ans.append(0)
        
    for i in range(k, N):
        if(A[i] < 0):
            q.append(i)
            
        if(len(q) > 0 and i - q[0] >= k):
            q.pop(0)
        
        if(len(q) > 0):
            ans.append(A[q[0]])
        else:
            ans.append(0)
            
    return ans

#{ 
#  Driver Code Starts
#Initial Template for Python 3

def main():

    T = int(input())

    while(T > 0):
        n = int(input())
        a = [int(x) for x in input().strip().split()]
        k = int(input())
        
        answer = printFirstNegativeInteger(a, n, k)
        for i in answer:
            print(i,end=" ")
        print()

        T -= 1


if __name__ == "__main__":
    main()


# } Driver Code Ends

File: Queue/Easy/first_negative_in_k_window.cpp
// https://practice.geeksforgeeks.org/problems/first-negative-integer-in-every-window-of-size-k3345/1#
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

vector<long long> printFirstNegativeInteger(long long int arr[],
                                             long long int n, long long int k);

// Driver program to test above functions
int main() {
    long long int t, i;
    cin >> t;
    while (t--) {
        long long int n;
        cin >> n;
        long long int arr[n];
        for (i = 0; i < n; i++) {
            cin >> arr[i];
        }
        long long int k;
        cin >> k;

        vector<long long> ans = printFirstNegativeInteger(arr, n, k);
        for (auto it : ans) cout << it << " ";
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

vector<long long> printFirstNegativeInteger(long long int A[],
                                             long long int N, long long int k) {
        queue<long long int> q;
        for(int i = 0; i < k; i++)
            if(A[i] < 0) q.push(i);
        
        vector<long long int> ans;
        if(q.size() > 0)
            ans.push_back(A[q.front()]);
        else
            ans.push_back(0);
        
        for(int i = k; i < N; i++)
        {
            if(!q.empty() && i - q.front() >= k) q.pop();
            if(A[i] < 0) q.push(i);
            
            if(!q.empty()) 
                ans.push_back(A[q.front()]);
            else 
                ans.push_back(0);
        }
        
        return ans;
 }